docs/internal/GeoffsRecoilAssessment.md

```diff
@@ -0,0 +1,176 @@
+- PRO: The default state of an atom with the reset() hook is convenient for clearing application state.
+- CON: Atoms are often arrays of items which means needing to look up items. This requires more undefined checks for cases where items are not found.
+- CON: There is additional "param baggage" to pass multiple atoms to different functions vs. passing the root of the application state or a child.
+- HAZARD: Care has to be taken to avoid updating atom state by mutating the current state.  Only a set will trigger change detection.
+
+## Selectors
+- PRO: Simple to shape data for React components. 
+- PRO: Like atoms they are small and can be built as an library of independent data shapes.
+- PRO: The dependency chain from one selector to other selectors or atoms is easy to follow.
+- PRO: Less ceremony compared to Redux's mapStateToProps.
+- UNKNOWN: Unsure if a selector would re-compute its value if no UI is subscribed to it currently but it had been subscribed before.
+- UNKNOWN: The writable selector is not well documented and unclear when to use.
+- HAZARD: Atom (and selector) key names need to be unique across the application. It can be easy to create an atom and selector with the same key.
+
+## useRecoilValue, useSetRecoilValue, useRecoilState hooks
+- PRO: The recoil state hooks cause rendering just like useState.
+- PRO: useRecoilValue and useSetRecoil value clearly denote the read-only and write-only scenarios.
+- PRO: No magic inspecting dot-notation usage like with MobX.  
+- HAZARD: Very easy to accidentally put a recoil hook within a condition. They need to be used as a block at the start of a component just like React hooks.
+
+## useRecoilCallback hook
+- PRO: Allows for creating methods that always get latest state even when called outside the context of a React component render. 
+- CON: It can be a bit of extra effort to get the Typescript types worked out for complex callbacks.
+- UNKNOWN: Unclear when to use the dependencies parameter vs. the getPromise to get latest state
+- UNKNOWN: Unclear when to use the synchronous loader vs the asynchronous getPromise.- UNKNOWN:? The callback documentation indicates is still possible this hook will change. This is likely due to the experimental portions relating to JS async threading.
+- HAZARD: Using getPromise within a callback requires the callback method be async, the return value be a promise, and usage outside UI callbacks use await.
+- HAZARD: Using set within a callback that is called multiple times in succession or called from other callbacks requires using updator function to ensure latest state.  Otherwise arrays may not be updated with the previous call and will be overridden - losing data.
+
+## RecoilRoot component
+- PRO: Simple to understand. It leverages the React context API, but more efficiently tracking atoms and selectors as keys in a dictionary.
+- HAZARD: The React component rendering RecoilRoot cannot use recoil hooks.
```

Comment:
This seems analogous to React having a "hazard" that anything outside the React root can't use React hooks. That's not a hazard, it's just how it works. On the other hand, I don't know what nesting RecoilRoots will do, so making complicated embedded structures is probably something to avoid (for now).

Reply:
Agreed. It seems like we wont find a usage of nesting recoil roots

Reply:
This could be something to explore for hosted UI plugins, or multiple tabs with different bot projects, etc. I don't think this will be something we do off the bat, but I could see if being an interesting technique to isolate certain parts of the application.

Reply:
@a-b-r-o-w-n I could see the multiple recoil roots being super effective for 3rd party plugins which need to have an isolated scope. 

For multi bot management I was thinking more on the line of an atom for BotProjects which is an array of BotId’s. Our LUFiles, LGFiles, Schemas individually are atoms containing arrays of  files with each object inside them referring to the BotID they  associate with. When a new tab  for a new project is opened or shifting between tabs I see the selector filtering out LUFiles and LGFiles atoms for that specific botID to refresh the view. What do u feel about this approach instead of going for swapping RecoilRoots based on the bot currently selected

Reply:
Yeah that is a reasonable approach.